File: 01jupyter/nb/source/17_linalg/problems/00_cg/cg.cpp
#include <cstdio>
#include <cstdlib>
#include <cmath>
#include <omp.h>

/* 共役勾配法 (CG) で 2次元ラプラシアン方程式 A x = b を解く (行列フリー)。
   未知数は n×n 格子上に並ぶので, CG の各ベクトル (x,b,r,p,Ap,xt) を Mat(n,n) で
   表す。行列ベクトル積 A p は 5点ステンシルで計算する (A は対称正定値)。
   内積や軸演算はベクトルとして中身 (.a) を一列に走査する。 */

/* 連続メモリの row-major 行列: A(i,j) = a[i*cols + j] */
struct Mat {
  long rows, cols;
  double * a;
  Mat(long r, long c)             : rows(r), cols(c), a(new double[r * c]) {}
  Mat(long r, long c, double * p) : rows(r), cols(c), a(p) {}
  Mat(Mat && o) noexcept : rows(o.rows), cols(o.cols), a(o.a) { o.a = nullptr; }
  ~Mat() { delete[] a; }
  Mat(const Mat &) = delete;
  Mat & operator=(const Mat &) = delete;
  double & operator()(long i, long j)       { return a[i * cols + j]; }
  double   operator()(long i, long j) const { return a[i * cols + j]; }
  long size() const { return rows * cols; }
  void zero() { for (long i = 0; i < rows * cols; i++) a[i] = 0.0; }
};

/* 状態を持たない乱数 (既知解の生成用): (seed,k) から [0,1)。 */
static inline double draw_rand01(long long seed, long long k) {
  const long long M = 2147483647LL;
  long long x = ((seed % M) * 2654435761LL + (k % M) + 1) % M;
  x = ((x ^ (x >> 16)) * 1812433253LL) % M;
  x = ((x ^ (x >> 13)) * 1664525LL)    % M;
  x =  (x ^ (x >> 16)) % M;
  return (double)x / (double)M;
}

/* 行列ベクトル積 y = A p。A は n×n 格子上の 2次元ラプラシアン (5点ステンシル,
   ディリクレ境界=0) で対称正定値。行列を保持せずステンシルの計算で済ませる。
   (A p)(i,j) = 4 p(i,j) - p(i-1,j) - p(i+1,j) - p(i,j-1) - p(i,j+1)  (領域外は 0)。 */
void matvec(const Mat & p, Mat & y) {
  long n = p.rows;
  // TODO: 格子点の二重ループを並列化する。
  // BEGIN ANSWER
#pragma omp parallel for collapse(2)
  // END ANSWER
  for (long i = 0; i < n; i++) {
    for (long j = 0; j < n; j++) {
      double v = 4.0 * p(i, j);
      if (i > 0)     v -= p(i - 1, j);
      if (i < n - 1) v -= p(i + 1, j);
      if (j > 0)     v -= p(i, j - 1);
      if (j < n - 1) v -= p(i, j + 1);
      y(i, j) = v;
    }
  }
}

/* 内積 a・b (ベクトルとみなして全要素の積和) */
double dot(const Mat & a, const Mat & b) {
  long N = a.size();
  double s = 0.0;
  // TODO: 内積の和を並列化する (reduction)。
  // BEGIN ANSWER
#pragma omp parallel for reduction(+:s)
  // END ANSWER
  for (long k = 0; k < N; k++) s += a.a[k] * b.a[k];
  return s;
}

int main(int argc, char ** argv) {
  int    n   = (argc > 1 ? atoi(argv[1]) : 128);    /* 格子の一辺 (未知数は N = n*n) */
  double tol = (argc > 2 ? atof(argv[2]) : 1e-8);
  long   N   = (long)n * n;
  int    maxit = 10 * n;

  Mat xt(n, n), b(n, n), x(n, n), r(n, n), p(n, n), Ap(n, n);

  for (long k = 0; k < N; k++) xt.a[k] = draw_rand01(k, 0);  /* 真の解をランダムに決め */
  matvec(xt, b);                                             /* b = A xt を作る */

  /* CG: x=0 から始めて A x = b を解く */
  for (long k = 0; k < N; k++) { x.a[k] = 0.0; r.a[k] = b.a[k]; p.a[k] = b.a[k]; }
  double rs = dot(r, r);

  int it;
  double t0 = omp_get_wtime();
  for (it = 0; it < maxit; it++) {
    matvec(p, Ap);
    double alpha = rs / dot(p, Ap);
    for (long k = 0; k < N; k++) { x.a[k] += alpha * p.a[k]; r.a[k] -= alpha * Ap.a[k]; }  /* (発展: ここも並列化可) */
    double rs_new = dot(r, r);
    if (sqrt(rs_new) < tol) { rs = rs_new; it++; break; }
    double beta = rs_new / rs;
    for (long k = 0; k < N; k++) p.a[k] = r.a[k] + beta * p.a[k];
    rs = rs_new;
  }
  double elapsed = omp_get_wtime() - t0;

  /* 検算: 求めた x が真の解 xt にどれだけ近いか */
  double err = 0.0;
  for (long k = 0; k < N; k++) { double e = fabs(x.a[k] - xt.a[k]); if (e > err) err = e; }
  printf("n=%d (N=%ld), iters=%d, 残差=%.2e, 解の誤差(max|x-xt|)=%.2e\n",
         n, N, it, sqrt(rs), err);
  printf("elapsed = %.3f sec\n", elapsed);
  return 0;
}

File: 01jupyter/nb/source/17_linalg/problems/00_cg/cg.f90
module cg_mod
contains
  ! 状態を持たない乱数 (既知解の生成用): (seed,k) から [0,1)。
  function draw_rand01(seed, k) result(u)
    integer(8), intent(in) :: seed, k
    real(8) :: u
    integer(8), parameter :: M = 2147483647_8
    integer(8) :: x
    x = modulo(modulo(seed, M) * 2654435761_8 + modulo(k, M) + 1_8, M)
    x = modulo(ieor(x, ishft(x, -16)) * 1812433253_8, M)
    x = modulo(ieor(x, ishft(x, -13)) * 1664525_8,    M)
    x = modulo(ieor(x, ishft(x, -16)), M)
    u = real(x, 8) / real(M, 8)
  end function draw_rand01

  ! 行列ベクトル積 y = A p。A は n×n 格子の 2次元ラプラシアン (5点ステンシル,
  ! ディリクレ境界=0) で対称正定値。行列を保持せずステンシルで計算する (行列フリー)。
  ! ベクトルは n×n のネイティブ2次元配列で表す。
  subroutine matvec(p, y)
    real(8), intent(in)  :: p(:,:)
    real(8), intent(out) :: y(:,:)
    integer :: i, j, n
    real(8) :: v
    n = size(p, 1)
    ! TODO: 格子点の二重ループを並列化する。
    ! BEGIN ANSWER
    !$omp parallel do collapse(2) private(v)
    ! END ANSWER
    do j = 1, n
       do i = 1, n
          v = 4.0d0 * p(i,j)
          if (i > 1) v = v - p(i-1,j)
          if (i < n) v = v - p(i+1,j)
          if (j > 1) v = v - p(i,j-1)
          if (j < n) v = v - p(i,j+1)
          y(i,j) = v
       end do
    end do
    ! BEGIN ANSWER
    !$omp end parallel do
    ! END ANSWER
  end subroutine matvec

  ! 内積 a・b (ベクトルとみなして全要素の積和)
  function dot(a, b) result(s)
    real(8), intent(in) :: a(:,:), b(:,:)
    real(8) :: s
    integer :: i, j
    s = 0.0d0
    ! TODO: 内積の和を並列化する (reduction)。
    ! BEGIN ANSWER
    !$omp parallel do collapse(2) reduction(+:s)
    ! END ANSWER
    do j = 1, size(a, 2)
       do i = 1, size(a, 1)
          s = s + a(i,j) * b(i,j)
       end do
    end do
    ! BEGIN ANSWER
    !$omp end parallel do
    ! END ANSWER
  end function dot
end module cg_mod

program cg
  use cg_mod
  use omp_lib
  character(len=32) :: arg
  integer :: n, maxit, it, i, j
  real(8) :: tol, rs, rs_new, alpha, beta, err, t0, elapsed
  real(8), allocatable :: xt(:,:), b(:,:), x(:,:), r(:,:), p(:,:), Ap(:,:)
  n = 128; tol = 1d-8
  if (command_argument_count() >= 1) then
     call get_command_argument(1, arg); read (arg, *) n
  end if
  if (command_argument_count() >= 2) then
     call get_command_argument(2, arg); read (arg, *) tol
  end if
  maxit = 10 * n

  allocate(xt(n,n), b(n,n), x(n,n), r(n,n), p(n,n), Ap(n,n))
  do i = 1, n
     do j = 1, n
        xt(i,j) = draw_rand01(int((i-1)*n + (j-1), 8), 0_8)   ! 真の解をランダムに決め
     end do
  end do
  call matvec(xt, b)                    ! b = A xt を作る

  ! CG: x=0 から始めて A x = b を解く
  x = 0.0d0; r = b; p = b
  rs = dot(r, r)

  t0 = omp_get_wtime()
  do it = 1, maxit
     call matvec(p, Ap)
     alpha = rs / dot(p, Ap)
     x = x + alpha * p; r = r - alpha * Ap     ! (発展: ここも並列化可)
     rs_new = dot(r, r)
     if (sqrt(rs_new) < tol) then
        rs = rs_new; exit
     end if
     beta = rs_new / rs
     p = r + beta * p
     rs = rs_new
  end do
  elapsed = omp_get_wtime() - t0

  ! 検算: 求めた x が真の解 xt にどれだけ近いか
  err = maxval(abs(x - xt))
  print "(a,i0,a,i0,a,i0,a,es9.2,a,es9.2)", &
       "n=", n, " (N=", n*n, "), iters=", min(it, maxit), &
       ", 残差=", sqrt(rs), ", 解の誤差(max|x-xt|)=", err
  print "(a,f0.3,a)", "elapsed = ", elapsed, " sec"
end program cg
